aiida_restapi/graphql/orm_factories.py

```diff
@@ -0,0 +1,242 @@
+    output = {}
+    for name, field in ORM_MAPPING[cls].__fields__.items():
+        if name in exclude_fields:
+            continue
+        gr_type = _type_mapping[field.type_]
+        output[name] = gr_type(description=field.field_info.description)
+    return output
+
+
+def field_names_from_orm(cls: Type[orm.Entity]) -> Set[str]:
+    """Extract the field names from an AIIDA ORM class."""
+    return set(get_model_from_orm(cls).__fields__.keys())
+
+
+def get_projection(
+    db_fields: Set[str], info: gr.ResolveInfo, is_link: bool = False
+) -> List[str]:
+    """Traverse the child AST, to work out what fields we should project.
+
+    Any fields found that are not database fields, are assumed to be joins.
+    If any joins are present, we always include "id", so they can be linked.
+
+    We fallback to "**" (all fields) if the selection set cannot be identified.
+    """
+    if is_link:
+        # TODO here we need to look deeper under the "node" field
+        return ["**"]
+    try:
+        selected = set(selected_field_names_naive(info.field_asts[0].selection_set))
+        fields = db_fields.intersection(selected)
+        joins = db_fields.difference(selected)
+        if joins:
+            fields.add("id")
+        return list(fields)
+    except NotImplementedError:
+        return ["**"]
+
+
+def single_cls_factory(
+    orm_cls: Type[orm.Entity], exclude_fields: Sequence[str] = ()
+) -> Type[gr.ObjectType]:
+    """Create a graphene class with standard fields/resolvers for querying a single AiiDA ORM entity."""
+    return type(
+        "AiidaOrmObjectType", (gr.ObjectType,), fields_from_orm(orm_cls, exclude_fields)
+    )
+
+
+EntitiesParentType = Optional[Dict[str, Any]]
+
+
+def create_query_path(
+    query: orm.QueryBuilder, parent: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Append parents to the query path, e.g. containing groups and incoming/outgoing nodes.
+
+    :param parent: data from the parent resolver
+    :returns: key-word arguments for the "leaf" path
+    """
+    leaf_kwargs: Dict[str, Any] = {}
+    if "group_id" in parent:
+        query.append(orm.Group, filters={"id": parent["group_id"]}, tag="group")
+        leaf_kwargs["with_group"] = "group"
+    if "edge_type" in parent:
+        query.append(
+            orm.nodes.Node,
+            filters={"id": parent["parent_id"]},
+            tag=parent["edge_type"],
+        )
+        leaf_kwargs[f'with_{parent["edge_type"]}'] = parent["edge_type"]
+        if parent.get("project_edge"):
+            leaf_kwargs["edge_tag"] = f'{parent["edge_type"]}_edge'
+            leaf_kwargs["edge_project"] = ["**"]
+    return leaf_kwargs
+
+
+def multirow_cls_factory(
+    entity_cls: Type[gr.ObjectType], orm_cls: Type[orm.Entity], name: str
+) -> Type[gr.ObjectType]:
+    """Create a graphene class with standard fields/resolvers for querying multiple rows of the same AiiDA ORM entity."""
```

Comment:
yes indeed. It was the only way I could think, to subclass a class containing dynamic `staticmethod` and `classattributes`, e.g. normally you might do something like:

```python
class A:
   b = 1
   def c(self):
      return self.b

class D(A):
    b = 2 
```

but how do you subclass?

```python
class A:
   b = 1
   @staticmethod
   def c():
      return b
```

certainly have a think if you see a better way to implement `multirow_cls_factory` and `single_cls_factory`, because it still feels a little iffy

Reply:
(although obviously it certainly appears to work)

Reply:
Is the `name` needed as an explicit input or could it be deduced?

If the input were only types (`entity_cls`, `orm_cls`), you could again use [generics](https://mypy.readthedocs.io/en/stable/generics.html)  here

Reply:
no `name` could be removed. Perhaps you could write the implementation you had in mind; I can quite picture how you intend to use generics just yet?